# This config file contains the configuration values that are used by the Panda pick and place task.
# FIXME: Environment issues with effort control and gravity compensation when hand attached (see https://github.com/frankaemika/franka_ros/issues/160#issuecomment-992776684).
panda_pick_and_place:
  ########################################
  # Control settings #####################
  ########################################
  control:
    direct_control: True # Directly control the panda robot by publishing on the controller command topics (FAST). When ``False`` the 'panda_gazebo' control services will be used (SLOWER).
    # ee_link: "panda_hand" # Uncomment to change end effector. Defaults: 'panda_hand' if gripper loaded, 'panda_link8' otherwise.
    load_gripper: True # Whether you want to load the gripper.
    # lock_gripper: False # Whether the gripper should be locked (i.e. not move).
    grasping: True # Whether the gripper should grasp objects. If true a max effort of 10N will be used when not specified otherwise.
    arm_wait: False # Wait for the arm control to complete before continuing to the next step.
    hand_wait: False # Wait for the hand control to complete before continuing to the next step.
    ee_control_coordinates: # Control variables used during ee-control.
      - "x"
      - "y"
      - "z"
      - "rx"
      - "ry"
      - "rz"
      - "rw"
    controlled_joints: # Joints that are controlled during joint position/effort control.
      - "panda_joint1"
      - "panda_joint2"
      - "panda_joint3"
      - "panda_joint4"
      - "panda_joint5"
      - "panda_joint6"
      - "panda_joint7"
      # NOTE: The current implementation only accepts gripper_width not individual finger joints.
      - "gripper_width"
      # NOTE: When uncommented the robot also tries to control the max gripper effort. Only used when 'gripper_width' is controlled.
      - "gripper_max_effort"
    locked_arm_joints: # Arm joints that are locked during control.
      # - "panda_joint1"
      # - "panda_joint2"
      # - "panda_joint3"
      # - "panda_joint4"
      # - "panda_joint5"
      # - "panda_joint6"
      # - "panda_joint7"

    # End-effector frame offset.
    # NOTE: Can be used to offset the end-effector pose returned by the robot. By default the transform between the end-effector link and
    # the 'panda_EE' frame is used.
    # WARNING: The 'ee_frame_offset' is not yet applied correctly when the control type is set to 'end_effector' (see #97).
    # ee_frame_offset:
    #   x: 0.0
    #   y: 0.0
    #   z: 0.0
    #   rx: 0.0
    #   ry: 0.0
    #   rz: 0.0
    #   rw: 1.0

  ########################################
  # Sampling settings ####################
  ########################################
  # Initial pose
  pose_sampling:
    visualize_init_pose_bounds: True # RViz pose sampling region visualization. Only works for 'pose_sampling_type' is 'ee_pose'.
    reset_init_pose: True # Reset the initial pose when resetting the simulation.
    random_init_pose: True # Use a random initial pose. The 'init_pose' field below will be used if set to `False`.
    randomize_first_episode: True # Also randomize the pose in the first episode.
    attempts: 10 # Maximum number of attempts for sampling a random initial pose within the set bounds.
    pose_sampling_type: "end_effector_pose" # From which set we should sample the initial pose. Options: 'end_effector_pose' and 'joint_positions'.
    moveit_control: False # Use MoveIt to set the initial pose. If `False` the Gazebo 'set_model_configuration' service will be used.

    # The initial robot pose (used when random is disabled).
    init_pose:
      x: 0.23
      y: 0.29
      z: 0.35
      rx: 0.78
      ry: 0.62
      rz: -0.0
      rw: 4.42
      panda_joint1: 0.0
      panda_joint2: 0.0
      panda_joint3: 0.0
      panda_joint4: -1.57079632679
      panda_joint5: 0.0
      panda_joint6: 1.57079632679
      panda_joint7: 0.785398163397
      gripper_width: 0.001

    # Initial ee pose position offset
    # NOTE: Ignored when 'random_init_pose' is `False` and 'pose_sampling_type' is `joint_positions`.
    offset:
      x: 0.0
      y: 0.0
      z: 0.0

    # Init pose sampling bounds
    # NOTE: Comment out if you don't want to clip the EE pose or joint positions during sampling.
    bounds:
      x_min: -0.6
      x_max: 0.6
      y_min: -0.6
      y_max: 0.6
      z_min: 0.0
      z_max: 1.2
      panda_joint1_min: -2.8973 # Min: -2.8973
      panda_joint1_max: 2.8973 # Max: 2.8973
      panda_joint2_min: -1.7628 # Min: -1.7628
      panda_joint2_max: 1.7628 # Max: 1.7628
      panda_joint3_min: -2.8973 # Min: -2.8973
      panda_joint3_max: 2.8973 # Max: 2.8973
      panda_joint4_min: -3.0718 # Min: -3.0718
      panda_joint4_max: -0.0698 # Max: -0.0698
      panda_joint5_min: -2.8973 # Min: -2.8973
      panda_joint5_max: 2.8973 # Max: 2.8973
      panda_joint6_min: -0.0175 # Min: -0.0175
      panda_joint6_max: 3.7525 # Max: 3.7525
      panda_joint7_min: -2.8973 # Min: -2.8973
      panda_joint7_max: 2.8973 # Max: 2.8973
      gripper_width_min: 0.0 # Min: 0.0
      gripper_width_max: 0.08 # Max: 0.08

  # Target sampling settings
  target_sampling:
    strategy: "global" # Options are: `global` and `fixed`.
    visualize_target: True # RViz target visualization.
    visualize_target_bounds: True # RViz target sampling region visualization.
    offset: # A additional offset applied to the target
      x: 0.0
      y: 0.0
      z: 0.0

    # Fixed target
    # NOTE: Used when strategy is set to `fixed`.
    fixed_target:
      x: 0.4
      y: 0.0
      z: 0.8

    # Random target sampling bounds
    bounds:
      global: # Relative to the world frame
        x_min: 0.1734215
        x_max: 0.6265785
        y_min: -0.3210245
        y_max: 0.3210245
        z_min: 0.419823 # NOTE: Starts from the platform
        z_max: 1.0

  # Object pose pose sampling settings
  object_sampling:
    strategy: "global" # Options are: `global` and `fixed`.
    distance_threshold: 0.1 # Threshold for determining whether new object pose is far enough from old obj pose.
    visualize_object_bounds: True # Turn object sampling region visualization on and off.

    # The pose of the grasp object
    # NOTE: Used when strategy is set to `fixed`.
    fixed_pose:
      x: 0.4
      y: 0.0
      z: 0.419088
      rx: 0.0
      ry: 0.0
      rz: 0.0
      rw: 1.0

    # Object sampling bounds
    bounds:
      x_min: 0.1734215
      x_max: 0.6265785
      y_min: -0.3210245
      y_max: 0.2610245

  ########################################
  # Training settings ####################
  ########################################
  training:
    object_name: "cube" # The name of the object the robot has to pick.
    platform_name: "platform" # The name of the platform present in the Gazebo environment.
    platform_size: [0.453157, 0.642049, 0.419823] # The size of the platform
    reward_type: "dense" # The reward type used in the reward function ("sparse" or "dense").
    target_hold: True # Hold the target position for N samples before done.
    hold_samples: 2 # The number of samples for which a agent should hold a target position.
    distance_threshold: 0.05 # The threshold for determining a target has been reached.
    collision_penalty: 1.0 # The penalty given for when the robot is in collision. Set to `0.0` if you don't want to use a collision penalty.

  ########################################
  # Environment settings #################
  ########################################
  environment:
    pause_after_step: True # Pause the simulation after each step.
    # The environment action space
    action_space:
      bounds:
        ee_pose:
          low:
            x: -1.3
            y: -1.3
            z: 0.0
            rx: 0
            ry: 0
            rz: 0
            rw: 0
          high:
            x: 1.3
            y: 1.3
            z: 1.3
            rx: 1
            ry: 1
            rz: 1
            rw: 1
        joint_positions: # NOTE: Limits were taken from the robot urdf file (see franka_description).
          low:
            panda_joint1: -2.8973
            panda_joint2: -1.7628
            panda_joint3: -2.8973
            panda_joint4: -3.0718
            panda_joint5: -2.8973
            panda_joint6: -0.0175
            panda_joint7: -2.8973
            gripper_width: 0.0
          high:
            panda_joint1: 2.8973
            panda_joint2: 1.7628
            panda_joint3: 2.8973
            panda_joint4: -0.0698
            panda_joint5: 2.8973
            panda_joint6: 3.7525
            panda_joint7: 2.8973
            gripper_width: 0.08
        joint_efforts:
          low:
            panda_joint1: -87
            panda_joint2: -87
            panda_joint3: -87
            panda_joint4: -87
            panda_joint5: -12
            panda_joint6: -12
            panda_joint7: -12
            gripper_max_effort: 0.0
          high:
            panda_joint1: 87
            panda_joint2: 87
            panda_joint3: 87
            panda_joint4: 87
            panda_joint5: 12
            panda_joint6: 12
            panda_joint7: 12
            gripper_max_effort: 140

  ########################################
  # Other pick and place settings ########
  ########################################
  physics: "ode" # The physics engine you want to use: dart|ode
  load_rviz: True # Whether RViz should be loaded.
  rviz_file: config/panda_pick_and_place.rviz # RViz config file. Path is relative to parent folder.
  load_gazebo_gui: False # Whether the gazebo GUI should be shown.
  log_reset: False # Whether we want to print a log statement when the world/simulation is reset.
  log_step_debug_info: False # Whether debug info about the step should be logged (i.e. reward, is_done, action etc.).
  roslaunch_log_to_console: False # Whether to write the log statements of the ROS launch files to the console.
##########################################
# Other settings #########################
##########################################
# NOTE: The settings below can be used to overwrite settings of other ROS nodes.
# panda_moveit_planner_server: # Scale the MoveIt control speed/ acceleration.
#   max_velocity_scaling_factor: 0.5
#   max_acceleration_scaling_factor: 0.5
